### PR TITLE
Sccarda/intrinsic fix monomorphization

### PR DESCRIPTION
**Pull request by ScottCarda-MS**
Prevent Intrinsic and certain Built In callables from being monomorphized or removed from the tree.

### CODE REVIEW

File: src/QsCompiler/Transformations/Monomorphization.cs [hunk 1]
```diff
@@ -157,31 +162,32 @@ public static QsCompilation Apply(QsCompilation compilation)
...
             }
 
             public class TransformationState
             {
-                public readonly ImmutableDictionary<NonNullable<string>, IEnumerable<QsCallable>> NamespaceCallables;
+                public readonly ILookup<NonNullable<string>, QsCallable> NamespaceCallables;
+                public readonly ImmutableHashSet<QsQualifiedName> ExemptCallableSet;
 
-                public TransformationState(ImmutableDictionary<NonNullable<string>, IEnumerable<QsCallable>> namespaceCallables)
+                public TransformationState(ILookup<NonNullable<string>, QsCallable> namespaceCallables, ImmutableHashSet<QsQualifiedName> exemptCallableSet)
```
Comment (bettinaheim) [inline src/QsCompiler/Transformations/Monomorphization.cs:177]: I like the use of the lookup!

File: src/QsCompiler/Transformations/Monomorphization.cs [hunk 2]
```diff
@@ -359,9 +378,13 @@ private class ExpressionKindTransformation : ExpressionKindTransformation<Transf
                             .Where(kvp => kvp.Key.Item1.Equals(global.Item))
                             .ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value);
 
-                        // Create a new identifier
-                        sym = SharedState.GetConcreteIdentifier(global, applicableParams);
-                        tArgs = QsNullable<ImmutableArray<ResolvedType>>.Null;
+                        // We want to skip over intrinsic callables. They will not be monomorphized.
```
Comment (bettinaheim) [inline src/QsCompiler/Transformations/Monomorphization.cs:381]: It's probably worth pointing this out in the doc comment of the transformation.

File: src/QsCompiler/Transformations/MonomorphizationValidation.cs [hunk 3]
```diff
@@ -16,26 +16,52 @@ public class ValidateMonomorphization : SyntaxTreeTransformation<ValidateMonomor
     {
```
Comment (bettinaheim) [inline src/QsCompiler/Transformations/MonomorphizationValidation.cs:16]: Looks like the validation is also complete! Great!

File: src/QsCompiler/Transformations/Monomorphization.cs (hunk 2 above)
Comment (ScottCarda-MS) [reply src/QsCompiler/Transformations/Monomorphization.cs:391]: I have added it to the doc comment.

**Pull request merged**